### PR TITLE
Port1.12/whatfix

### LINKED ISSUE #1504: Players cannot log in after server restart
1.12 Minecolonies version

version 0.8.4118 and 0.8.4191

###  Expected behavior

Restart the server.

### Actual behaviour

Restarts, and kicks people upon login via "Internal Exception: java.io.IOException forcibly closed by remote host error" type error.

Backend looks like;
[image](https://i.imgur.com/CbnJ6GZ.png)
`[12:35:26] [User Authenticator #3/INFO]: UUID of player PotatoOrgy is 327fd2d5-b994-4223-845b-1cde6779df92
[12:35:39] [Server thread/INFO]: PotatoOrgy[/--] logged in with entity id 664 at (-1277.8143034057182, 85.0, 6.643822882801938)
[12:35:39] [Server thread/INFO]: PotatoOrgy joined the game
[12:35:39] [Server thread/WARN]: Failed to handle packet for /--
java.lang.NullPointerException: null
	at com.minecolonies.coremod.util.ServerUtils.getPlayerFromUUID(ServerUtils.java:143) ~[ServerUtils.class:?]
	at com.minecolonies.coremod.util.ServerUtils.getPlayerFromPermPlayer(ServerUtils.java:124) ~[ServerUtils.class:?]
	at com.minecolonies.coremod.util.ServerUtils.getPlayersFromPermPlayer(ServerUtils.java:103) ~[ServerUtils.class:?]
	at com.minecolonies.coremod.util.AchievementUtils.syncAchievements(AchievementUtils.java:37) ~[AchievementUtils.class:?]
	at java.lang.Iterable.forEach(Iterable.java:75) ~[?:1.8.0_144]
	at com.minecolonies.coremod.colony.ColonyManager.syncAllColoniesAchievements(ColonyManager.java:231) ~[ColonyManager.class:?]
	at com.minecolonies.coremod.event.FMLEventHandler.onPlayerLogin(FMLEventHandler.java:83) ~[FMLEventHandler.class:?]
	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler_231_FMLEventHandler_onPlayerLogin_PlayerLoggedInEvent.invoke(.dynamic) ~[?:?]
	at net.minecraftforge.fml.common.eventhandler.ASMEventHandler.invoke(ASMEventHandler.java:90) ~[ASMEventHandler.class:?]
	at net.minecraftforge.fml.common.eventhandler.EventBus.post(EventBus.java:179) ~[EventBus.class:?]
	at net.minecraftforge.fml.common.FMLCommonHandler.firePlayerLoggedIn(FMLCommonHandler.java:573) ~[FMLCommonHandler.class:?]
	at net.minecraft.server.management.PlayerList.initializeConnectionToPlayer(PlayerList.java:218) ~[pl.class:?]
	at net.minecraftforge.fml.common.network.handshake.NetworkDispatcher.completeServerSideConnection(NetworkDispatcher.java:256) ~[NetworkDispatcher.class:?]
	at net.minecraftforge.fml.common.network.handshake.NetworkDispatcher.access$100(NetworkDispatcher.java:72) ~[NetworkDispatcher.class:?]
	at net.minecraftforge.fml.common.network.handshake.NetworkDispatcher$1.func_73660_a(NetworkDispatcher.java:205) ~[NetworkDispatcher$1.class:?]
	at net.minecraft.network.NetworkManager.func_74428_b(NetworkManager.java:285) ~[gw.class:?]
	at net.minecraft.network.NetworkSystem.func_151269_c(NetworkSystem.java:179) [oz.class:?]
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:788) [MinecraftServer.class:?]
	at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:396) [nz.class:?]
	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:666) [MinecraftServer.class:?]
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:524) [MinecraftServer.class:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
[12:35:44] [Server thread/INFO]: PotatoOrgy lost connection: Internal server error
[12:35:44] [Server thread/INFO]: PotatoOrgy left the game`

### Steps to reproduce the problem

1. Restart the server
2. Problem
3. ...
4. Profit

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #1504

security measures in tough times

Fix despawning

correctly spawn barbarians

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/event/ClientEventHandler.java [hunk 1]
```diff
@@ -101,7 +101,7 @@ public void renderWorldLastEvent(@NotNull final RenderWorldLastEvent event)
             }
 
             final ColonyView colony = ColonyManager.getClosestColonyView(world, player.getPosition());
-            if(colony != null && colony.getPermissions().hasPermission(player, Action.ACCESS_HUTS))
+            if(colony != null && player != null && colony.getPermissions().hasPermission(player, Action.ACCESS_HUTS))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/event/ClientEventHandler.java:104]: MAJOR Remove this expression which always evaluates to "true" [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2589)

File: src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java [hunk 2]
```diff
@@ -938,7 +938,7 @@ public boolean hasPermission(final UUID id, @NotNull final Action action)
         public boolean hasPermission(final Rank rank, @NotNull final Action action)
         {
             return (rank == Rank.OWNER && action != Action.GUARDS_ATTACK)
-                     || Utils.testFlag(permissions.get(rank), action.getFlag());
+                     || (permissions != null && action != null && Utils.testFlag(permissions.get(rank), action.getFlag()));
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java:941]: MAJOR Remove this expression which always evaluates to "true" [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2589)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 66 issues
* ![MAJOR][MAJOR] 22 major
* ![MINOR][MINOR] 41 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java](https://github.com/Minecolonies/minecolonies/blob/030e05cab4aff936b801df0ced9c9fcdaf22f123/src/main/java/com/minecolonies/coremod/colony/Colony.java): This file has 1.355 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [Colony.java#L55](https://github.com/Minecolonies/minecolonies/blob/030e05cab4aff936b801df0ced9c9fcdaf22f123/src/main/java/com/minecolonies/coremod/colony/Colony.java#L55): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 49 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [Colony.java#L55](https://github.com/Minecolonies/minecolonies/blob/030e05cab4aff936b801df0ced9c9fcdaf22f123/src/main/java/com/minecolonies/coremod/colony/Colony.java#L55): class "Colony" has 94 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java#L230](https://github.com/Minecolonies/minecolonies/blob/030e05cab4aff936b801df0ced9c9fcdaf22f123/src/main/java/com/minecolonies/coremod/colony/Colony.java#L230): Assign this magic number 5 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L355](https://github.com/Minecolonies/minecolonies/blob/030e05cab4aff936b801df0ced9c9fcdaf22f123/src/main/java/com/minecolonies/coremod/colony/Colony.java#L355): Remove this useless assignment to local variable "achievementKey". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Colony.java#L514](https://github.com/Minecolonies/minecolonies/blob/030e05cab4aff936b801df0ced9c9fcdaf22f123/src/main/java/com/minecolonies/coremod/colony/Colony.java#L514): Remove this useless assignment to local variable "achievementCompound". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Colony.java#L635](https://github.com/Minecolonies/minecolonies/blob/030e05cab4aff936b801df0ced9c9fcdaf22f123/src/main/java/com/minecolonies/coremod/colony/Colony.java#L635): Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
1. ![MAJOR][MAJOR] [Colony.java#L1084](https://github.com/Minecolonies/minecolonies/blob/030e05cab4aff936b801df0ced9c9fcdaf22f123/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1084): The Cyclomatic Complexity of this method "onWorldTick" is 19 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [Colony.java#L1266](https://github.com/Minecolonies/minecolonies/blob/030e05cab4aff936b801df0ced9c9fcdaf22f123/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1266): Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1268](https://github.com/Minecolonies/minecolonies/blob/030e05cab4aff936b801df0ced9c9fcdaf22f123/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1268): Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 66 issues
* ![MAJOR][MAJOR] 22 major
* ![MINOR][MINOR] 41 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java](https://github.com/Minecolonies/minecolonies/blob/3809b7dd43f90cde7158ba79f5e277dd1b4e2406/src/main/java/com/minecolonies/coremod/colony/Colony.java): This file has 1.355 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [Colony.java#L55](https://github.com/Minecolonies/minecolonies/blob/3809b7dd43f90cde7158ba79f5e277dd1b4e2406/src/main/java/com/minecolonies/coremod/colony/Colony.java#L55): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 49 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [Colony.java#L55](https://github.com/Minecolonies/minecolonies/blob/3809b7dd43f90cde7158ba79f5e277dd1b4e2406/src/main/java/com/minecolonies/coremod/colony/Colony.java#L55): class "Colony" has 94 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java#L230](https://github.com/Minecolonies/minecolonies/blob/3809b7dd43f90cde7158ba79f5e277dd1b4e2406/src/main/java/com/minecolonies/coremod/colony/Colony.java#L230): Assign this magic number 5 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L355](https://github.com/Minecolonies/minecolonies/blob/3809b7dd43f90cde7158ba79f5e277dd1b4e2406/src/main/java/com/minecolonies/coremod/colony/Colony.java#L355): Remove this useless assignment to local variable "achievementKey". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Colony.java#L514](https://github.com/Minecolonies/minecolonies/blob/3809b7dd43f90cde7158ba79f5e277dd1b4e2406/src/main/java/com/minecolonies/coremod/colony/Colony.java#L514): Remove this useless assignment to local variable "achievementCompound". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Colony.java#L635](https://github.com/Minecolonies/minecolonies/blob/3809b7dd43f90cde7158ba79f5e277dd1b4e2406/src/main/java/com/minecolonies/coremod/colony/Colony.java#L635): Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
1. ![MAJOR][MAJOR] [Colony.java#L1084](https://github.com/Minecolonies/minecolonies/blob/3809b7dd43f90cde7158ba79f5e277dd1b4e2406/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1084): The Cyclomatic Complexity of this method "onWorldTick" is 19 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [Colony.java#L1266](https://github.com/Minecolonies/minecolonies/blob/3809b7dd43f90cde7158ba79f5e277dd1b4e2406/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1266): Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1268](https://github.com/Minecolonies/minecolonies/blob/3809b7dd43f90cde7158ba79f5e277dd1b4e2406/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1268): Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — APPROVED** [review]

File: src/main/java/com/minecolonies/coremod/event/FMLEventHandler.java [hunk 3]
```diff
@@ -80,7 +80,7 @@ public void onPlayerLogin(@NotNull final PlayerEvent.PlayerLoggedInEvent event)
         {
             MineColonies.getNetwork().sendTo(new ServerUUIDMessage(), (EntityPlayerMP) event.player);
             MineColonies.getNetwork().sendTo(new ColonyStylesMessage(), (EntityPlayerMP) event.player);
-            ColonyManager.syncAllColoniesAchievements();
+            //ColonyManager.syncAllColoniesAchievements();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/event/FMLEventHandler.java:83]: MINOR This block of commented-out lines of code should be removed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 71 issues
* ![MAJOR][MAJOR] 22 major
* ![MINOR][MINOR] 46 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [Colony.java](https://github.com/Minecolonies/minecolonies/blob/4beaa04407221578e40bbb5d2d15143c59425d9e/src/main/java/com/minecolonies/coremod/colony/Colony.java): This file has 1.355 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [Colony.java#L55](https://github.com/Minecolonies/minecolonies/blob/4beaa04407221578e40bbb5d2d15143c59425d9e/src/main/java/com/minecolonies/coremod/colony/Colony.java#L55): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 49 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [Colony.java#L55](https://github.com/Minecolonies/minecolonies/blob/4beaa04407221578e40bbb5d2d15143c59425d9e/src/main/java/com/minecolonies/coremod/colony/Colony.java#L55): class "Colony" has 94 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java#L230](https://github.com/Minecolonies/minecolonies/blob/4beaa04407221578e40bbb5d2d15143c59425d9e/src/main/java/com/minecolonies/coremod/colony/Colony.java#L230): Assign this magic number 5 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L355](https://github.com/Minecolonies/minecolonies/blob/4beaa04407221578e40bbb5d2d15143c59425d9e/src/main/java/com/minecolonies/coremod/colony/Colony.java#L355): Remove this useless assignment to local variable "achievementKey". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Colony.java#L514](https://github.com/Minecolonies/minecolonies/blob/4beaa04407221578e40bbb5d2d15143c59425d9e/src/main/java/com/minecolonies/coremod/colony/Colony.java#L514): Remove this useless assignment to local variable "achievementCompound". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [Colony.java#L635](https://github.com/Minecolonies/minecolonies/blob/4beaa04407221578e40bbb5d2d15143c59425d9e/src/main/java/com/minecolonies/coremod/colony/Colony.java#L635): Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
1. ![MAJOR][MAJOR] [Colony.java#L1084](https://github.com/Minecolonies/minecolonies/blob/4beaa04407221578e40bbb5d2d15143c59425d9e/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1084): The Cyclomatic Complexity of this method "onWorldTick" is 19 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [Colony.java#L1266](https://github.com/Minecolonies/minecolonies/blob/4beaa04407221578e40bbb5d2d15143c59425d9e/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1266): Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [Colony.java#L1268](https://github.com/Minecolonies/minecolonies/blob/4beaa04407221578e40bbb5d2d15143c59425d9e/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1268): Assign this magic number 16 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**